### ISSUE TITLE
MLFlow integration with Lightning + Neuron

### ISSUE THREAD

**Issue by BrianF-tessera** [aws-neuron/aws-neuron-sdk#810]
The setup:

Basic MNIST model as proof of concept using pytorch lightning with xla libraries on top of the Neuron DLAMI. Training on 2 Neuron cores (trn1.2xlarge) works well and successfully calls the Neuron complier during training.

The issues:

Logging parameters to MLFlow works on 1 Neuron core, but when using multiple cores fails to save parameters (no error thrown, just no readout in the MLFlow server) and when attempting to log/save the model, even with one Neuron core there's a failure with python's stack trace just pointing to the line of code that logs the model. 

Unsurprisingly the mlflow autologger also fails, but that's fine since Neuron is a far from mature/integrated system.

**Comment by aws-rhsoln** [comment]
Thank you for reporting the issue. We are building on adding support for PTL integration end to end in our upcoming releases and we will look into this issue. Can you provide minimal example on how to reproduce the issue (eg. training scripts, packages installed). 

We have released beta support for PTL with Neuronx-Distributed library to help customers use PTL for distributed training. Please refer to some docs for details about the the integration: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/neuronx-distributed/ptl_developer_guide.html#ptl-developer-guide
Sample example: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/neuronx-distributed/tutorials/training_llama2_7b_ptl.html#llama2-7b-tp-zero1-ptl-tutorial

**Comment by BrianF-tessera** [comment]
`import os
import torch
import torch.nn as nn
import torch
import lightning.pytorch as pl
from lightning.pytorch.loggers import MLFlowLogger
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST
from torchvision import transforms
from torchmetrics import Accuracy
import torch_xla.core.xla_model as xm
import mlflow


class MNISTClassifier(pl.LightningModule):
    def __init__(self):
        super(MNISTClassifier, self).__init__()
        self.accuracy = Accuracy(task="multiclass", num_classes=10, top_k=1)
        self.layer_1_size = 128
        self.layer_2_size = 256
        self.lr = .01

        # mnist images are (1, 28, 28) (channels, width, height)
        self.layer_1 = torch.nn.Linear(28 * 28, self.layer_1_size)
        self.layer_2 = torch.nn.Linear(self.layer_1_size, self.layer_2_size)
        self.layer_3 = torch.nn.Linear(self.layer_2_size, 10)
        self.eval_loss = []
        self.eval_accuracy = []

    def cross_entropy_loss(self, logits, labels):
        return F.nll_loss(logits, labels)

    def forward(self, x):
        batch_size, channels, width, height = x.size()
        x = x.view(batch_size, -1)

        x = self.layer_1(x)
        x = torch.relu(x)

        x = self.layer_2(x)
        x = torch.relu(x)

        x = self.layer_3(x)
        x = torch.log_softmax(x, dim=1)

        return x

    def training_step(self, train_batch, batch_idx):
        x, y = train_batch
        logits = self.forward(x)
        loss = self.cross_entropy_loss(logits, y)
        accuracy = self.accuracy(logits, y)

        self.log("ptl/train_loss", loss)
        self.log("ptl/train_accuracy", accuracy)
        return loss

    def validation_step(self, val_batch, batch_idx):
        x, y = val_batch
        logits = self.forward(x)
        loss = self.cross_entropy_loss(logits, y)
        accuracy = self.accuracy(logits, y)
        self.eval_loss.append(loss)
        self.eval_accuracy.append(accuracy)
        return {"val_loss": loss, "val_accuracy": accuracy}

    def on_validation_epoch_end(self):
        avg_loss = torch.stack(self.eval_loss).mean()
        avg_acc = torch.stack(self.eval_accuracy).mean()
        self.log("ptl/val_loss", avg_loss, sync_dist=True)
        self.log("ptl/val_accuracy", avg_acc, sync_dist=True)
        self.eval_loss.clear()
        self.eval_accuracy.clear()

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer


def main():
    print(mlflow.__version__)
    mlflow.set_tracking_uri('mlflow.ml.tesseratx.net')  # set up connection
    mlflow.set_experiment('MNIST_xla_demo')  # set the experiment
    # data
    mnist_train = MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor())
    train_sampler = torch.utils.data.distributed.DistributedSampler(
        mnist_train,
        num_replicas=xm.xrt_world_size(),
        rank=xm.get_ordinal(),
        shuffle=True
    )
    mnist_train = DataLoader(mnist_train, sampler=train_sampler, batch_size=512, num_workers=4)
    mnist_val = MNIST(os.getcwd(), download=True, transform=transforms.ToTensor())
    val_sampler = torch.utils.data.distributed.DistributedSampler(
        mnist_val,
        num_replicas=xm.xrt_world_size(),
        rank=xm.get_ordinal(),
        shuffle=True
    )
    mnist_val = DataLoader(mnist_val, sampler=val_sampler, batch_size=512, num_workers=4)

    # model
    mnist_model = MNISTClassifier()
    mlf_logger = MLFlowLogger(experiment_name="MNIST_xla_demo",
                              tracking_uri="mlflow.ml.tesseratx.net",
                              log_model=True)
    trainer = pl.Trainer(devices=2, precision='bf16-true', max_epochs=200, check_val_every_n_epoch=25, loggers=[mlf_logger])
    trainer.fit(mnist_model, mnist_train, mnist_val)


if __name__ == "__main__":
    main()
`

This is a minimum viable example that I used to demonstrate who denovo training/model development and design would look like. I can work through tutorials and they run fine. Problems only come when interacting with the rest of our toolchain
